src/coreclr/vm/object.cpp

```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+    {
+        struct
+        {
+            StackTraceArray newStackTrace;
+            PTRARRAYREF newKeepAliveArray = NULL;
+        } gc;
+
+        GCPROTECT_BEGIN(gc);
+
+        // When the stack trace was created by other thread than the current one, we create a copy of both the stack trace and the keepAlive arrays to make sure
+        // they are not changing while the caller is accessing them.
+        wasCreatedByForeignThread = true;
+        gc.newStackTrace.Allocate(stackTrace.Capacity());
+        gc.newStackTrace.CopyDataFrom(stackTrace);
+        stackTrace.Set(gc.newStackTrace.Get());
 
-    StackTraceArray temp(_stackTrace);
-    stackTrace.Swap(temp);
+        uint32_t keepAliveArrayCapacity = ((*outKeepAliveArray) == NULL) ? 0 : (*outKeepAliveArray)->GetNumComponents();
+
+        uint32_t j = 0;
+        uint32_t count = stackTrace.Size();
+        for (uint32_t i = 0; i < count; i++)
+        {
+            if (stackTrace[i].flags & STEF_KEEPALIVE)
+            {
+                OBJECTREF keepAliveObject = NULL;
+                if ((j + 1) < keepAliveArrayCapacity)
+                {
+                    keepAliveObject = (*outKeepAliveArray)->GetAt(j + 1);
+                }
+                if (keepAliveObject == NULL)
+                {
+                    // Trim the stack trace at a point where a dynamic or collectible method is found without a corresponding keepAlive object.
+                    stackTrace.SetSize(i);
+                    stackTrace.SetKeepAliveItemsCount(j);
+                    _ASSERTE(stackTrace.ComputeKeepAliveItemsCount() == j);
+                    break;
+                }
+                j++;
+            }
+        }
+
+        if (keepAliveArrayCapacity != 0)
+        {
+            gc.newKeepAliveArray = (PTRARRAYREF)AllocateObjectArray(static_cast<DWORD>(keepAliveArrayCapacity), g_pObjectClass);
+            _ASSERTE((*outKeepAliveArray) != NULL);
+            memmoveGCRefs(gc.newKeepAliveArray->GetDataPtr() + 1,
+                          (*outKeepAliveArray)->GetDataPtr() + 1,
+                          (keepAliveArrayCapacity - 1) * sizeof(Object *));
+            gc.newKeepAliveArray->SetAt(0, stackTrace.Get());
+            *outKeepAliveArray = gc.newKeepAliveArray;
+        }
+        else
+        {
+            *outKeepAliveArray = NULL;
+        }
+        GCPROTECT_END();
+    }
+#endif // DACCESS_COMPILE            
+ 
+    return wasCreatedByForeignThread;
+}
 
-    if (outDynamicMethodArray != NULL)
+// Get the stack trace and the dynamic method array from the stack trace object. 
+// If the stack trace was created by another thread, it returns clones of both arrays.
+/* static */
+void ExceptionObject::GetStackTraceParts(OBJECTREF stackTraceObj, StackTraceArray & stackTrace, PTRARRAYREF * outKeepAliveArray /*= NULL*/)
+{
+    CONTRACTL
     {
-        *outDynamicMethodArray = _dynamicMethods;
+        GC_TRIGGERS;
+        THROWS;
```

Comment:
```suggestion
        GC_NOTRIGGER;
        NOTHROW;
```
This does not seem to trigger GC or throw

Reply:
Thanks for spotting this, it used to have an allocation, now it doesn't.

Reply:
removed